File: 221.最大正方形/solution.py
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        m, n = len(matrix), len(matrix[0])
        f = [[0 for _ in range(n)] for _ in range(m)]
        length = 0
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == '1':
                    if i == 0 or j == 0:
                        f[i][j] = 1
                    else:
                        f[i][j] = min(f[i][j - 1], f[i - 1][j], f[i - 1][j - 1]) + 1
                length = max(length, f[i][j])
        
        return length * length

File: 221.最大正方形/solution.cpp
# include "..\leetcode.h"

class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        // f[i, j]以matrix[i, j]做右下角的最大正方形边长
        vector<vector<int>> f(m, vector<int>(n, 0));
        int len = 0;
        for(int i = 0; i < m; i++) {
            for(int j = 0; j < n; j++) {
                if(matrix[i][j] == '1') {
                    if(i == 0 || j == 0) {
                        f[i][j] = 1;
                    }else {
                        f[i][j] = min(min(f[i - 1][j], f[i][j - 1]), f[i - 1][j - 1]) + 1;
                    }
                }
                len = max(len, f[i][j]);
            }
        }
        return len * len;
    }
};